applications/services/input/input.c

```diff
@@ -75,95 +72,185 @@ const char* input_get_type_name(InputType type) {
+
+        key_sequence->type = InputTypeRepeat;
+        break;
+    case InputTypeRelease:
+        if(key_sequence->press_counter < INPUT_SRV_LONG_PRESS_COUNTS) {
+            event.sequence_source = INPUT_SEQUENCE_SOURCE_HARDWARE;
+            event.sequence_counter = key_sequence->sequence_counter;
+            event.key = key_sequence->key;
+            event.type = InputTypeShort;
+            input_send(key_sequence->event_pubsub, &event);
         }
+        event.sequence_source = INPUT_SEQUENCE_SOURCE_HARDWARE;
+        event.sequence_counter = key_sequence->sequence_counter;
+        event.key = key_sequence->key;
+        event.type = InputTypeRelease;
+        input_send(key_sequence->event_pubsub, &event);
+
+        furi_event_loop_timer_stop(key_sequence->timer);
+
+        key_sequence->type = InputTypeRelease;
+        break;
+
+    default:
+        break;
     }
+}
+static void input_sequence_timer_callback(void* context) {
+    furi_assert(context);
+    InputSrvKeySequence* key_sequence = context;
+    InputEvent event;
+
+    if(key_sequence->press_counter == INPUT_SRV_LONG_PRESS_COUNTS) {
+        event.sequence_source = INPUT_SEQUENCE_SOURCE_HARDWARE;
+        event.sequence_counter = key_sequence->sequence_counter;
+        event.key = key_sequence->key;
+        event.type = InputTypeLong;
+        input_send(key_sequence->event_pubsub, &event);
+    } else if(key_sequence->press_counter > INPUT_SRV_LONG_PRESS_COUNTS) {
+        event.sequence_source = INPUT_SEQUENCE_SOURCE_HARDWARE;
+        event.sequence_counter = key_sequence->sequence_counter;
+        event.key = key_sequence->key;
+        event.type = InputTypeRepeat;
+        input_send(key_sequence->event_pubsub, &event);
+    }
+
+    key_sequence->press_counter++;
+}
+
+int32_t input_srv(void* p) {
+    UNUSED(p);
+    InputSrv* instance = malloc(sizeof(InputSrv));
+    instance->event_pubsub = furi_pubsub_alloc();
+    furi_record_create(RECORD_INPUT_EVENTS, instance->event_pubsub);
+
+    instance->input_semaphore = furi_semaphore_alloc(1, 0);
+    instance->event_loop = furi_event_loop_alloc();
+    instance->debounce_timer = furi_event_loop_timer_alloc(
+        instance->event_loop,
+        input_debounce_timer_callback,
+        FuriEventLoopTimerTypePeriodic,
+        instance);
+
+    instance->key_state = malloc(sizeof(InputSrvKeyState) * input_pins_count);
+    for(size_t i = 0; i < input_pins_count; i++) {
+        furi_hal_gpio_add_int_callback(input_pins[i].gpio, input_isr_key, instance);
+        instance->key_state[i].pin = &input_pins[i];
+        instance->key_state[i].state = GPIO_Read(instance->key_state[i]);
+        instance->sequence_counter = 0;
+    }
+
+    furi_event_loop_subscribe_semaphore(
+        instance->event_loop,
+        instance->input_semaphore,
+        FuriEventLoopEventIn,
+        input_semaphore_callback,
+        instance);
+
+    instance->key_sequence = malloc(sizeof(InputSrvKeySequence) * input_pins_count);
```

Comment:
All those allocations could easily land on the stack (copying my remark from the previous PR).

Reply:
no, if you put everything on the stack and turn on debugging, then the 1 KB stack allocated to the process is sometimes not enough and everything crashes with special effects

Reply:
Those structures are tiny - so wouldn't this mean that even without them on the stack, 1KB is pushing the limit very close?

Reply:
all structures on the stack will take about 220 bytes.

provided that there are only 6 buttons, when the number of buttons changes, the size of the structure will increase and if the structure is placed on the stack, the implementation ceases to be universal.

Reply:
at the expense of 1kb on the stack it is enough even with debugging enabled, without it about 500 bytes are used per process

Reply:
In this case the keys could stay on the heap (although maybe put them on the service heap?) while `instance` goes to the stack?

Reply:
Well then what's the point of chopping it into pieces?

Reply:
Currently it's chopped into pieces already, and occupies 3 heap allocations - so by moving you save one allocation, which is better than nothing.

Reply:
Keep in mind that services are allocated in special memory region and it's quite small. Increasing stack size may bring another set of problems

Reply:
Sure, in this case the key arrays could stay on the user heap, but the inputsrv is so small there's no reason not to put it on the stack. Before this PR it was just a bunch of local variables.